### PR TITLE
[Explicit Module Builds] Only specify '-fmodule-map-file' for bridging header Clang module dependencies

### PR DESCRIPTION
**Pull request by artemcm**
Relying on the corresponding field in the '-explicit-swift-module-map-file' provided by the driver.

- Only bridging headers require a module map because that's what aids header include resolution. With lazy module loading today, '.modulemap' parsing which happens when instantiating Clang is responsible for associating headers with modules. Then upon encountering a header include inside the bridging header the compiler knows which module corresponds to said header and is then able to load explicitly-provided PCM for that module. For all other module dependencies, they are only ever queried by-name from Swift, so '.modulemap' parsing is not necessary.

- [Dependency Scanning] Add libSwiftScan entrypoint to query binary module dependencies' header inputs' module dependencies

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci smoke test

**Review by cachemeifyoucan — APPROVED** [review]

**Comment by artemcm** [comment]
@cachemeifyoucan sorry, I've expanded the scope of this PR, because I want these changes to land in-tandem on the compiler. Could you please take another look?

**Comment by artemcm** [comment]
@swift-ci test

File: include/swift/AST/SearchPathOptions.h [hunk 1]
```diff
@@ -476,7 +476,7 @@ class SearchPathOptions {
   std::vector<std::string> CandidateCompiledModules;
 
   /// A map of explicit Swift module information.
-  std::string ExplicitSwiftModuleMap;
+  std::string ExplicitSwiftModuleMapPath;
```
Comment (cachemeifyoucan) [inline include/swift/AST/SearchPathOptions.h:479]: The caching build doesn't really take a path here (but a CASID). Maybe it wasn't a good idea to reuse this flag for caching build for the first place.

**Review by cachemeifyoucan — APPROVED** [review]
LGTM in general. I always wondering if swift scanner should write the file or content pass to `explicit-swift-module-map-file` directly so there is no need to add more APIs to traffic more and more information to driver so driver can write this file.

**Pull request merged**